File: Algorithms/DLX/F_Sharp/dlx.fs

open System
open System.IO
open System.Diagnostics

// Types
[<AllowNullLiteral>]
type DlxNode() =
    member val Up: DlxNode = null with get, set
    member val Down: DlxNode = null with get, set
    member val Left: DlxNode = null with get, set
    member val Right: DlxNode = null with get, set
    member val Column: DlxColumn = null with get, set
    member val RowId: int = -1 with get, set

and [<AllowNullLiteral>]
    DlxColumn() =
    inherit DlxNode()
    member val Size: int = 0 with get, set
    member val Name: string = "" with get, set
    new(n: string) as this = 
        DlxColumn() then
        this.Name <- n
        this.Column <- this

type RowInfo = {
    Row: int
    Col: int
    Num: int
}

// Global state
let mutable dlxIterations = 0
let mutable root: DlxColumn = null
let mutable columns: DlxColumn[] = null
let mutable nodes: DlxNode[] = null
let mutable nodeCount = 0
let mutable rowInfo: RowInfo[] = null
let mutable rowStarts: DlxNode[] = null
let mutable puzzle: int[,] = Array2D.create 9 9 0
let mutable solutionGrid: int[,] = Array2D.create 9 9 0

let maxNodes = 729 * 4

// --- DLX Logic ---

let initDlxMatrix () =
    root <- new DlxColumn("root")
    root.Left <- root
    root.Right <- root
    root.Up <- root
    root.Down <- root
    
    columns <- Array.zeroCreate 324
    for i in 0 .. 323 do
        let col = new DlxColumn(sprintf "C%d" i)
        columns.[i] <- col
        col.Up <- col
        col.Down <- col
        
        // Link to header
        col.Left <- root.Left
        col.Right <- root
        root.Left.Right <- col
        root.Left <- col

    nodes <- Array.zeroCreate maxNodes
    for i in 0 .. maxNodes - 1 do
        nodes.[i] <- new DlxNode()
    
    nodeCount <- 0
    rowInfo <- Array.zeroCreate 729
    rowStarts <- Array.zeroCreate 729

let addNode (col: DlxColumn) (rowId: int) =
    if nodeCount >= maxNodes then failwith "Max nodes exceeded"
    
    let node = nodes.[nodeCount]
    nodeCount <- nodeCount + 1
    
    node.Column <- col
    node.RowId <- rowId
    
    // Insert at bottom
    node.Down <- col
    node.Up <- col.Up
    col.Up.Down <- node
    col.Up <- node
    col.Size <- col.Size + 1
    
    node

let getPositionCol r c = r * 9 + c
let getRowCol r n = 81 + r * 9 + (n - 1)
let getColCol c n = 162 + c * 9 + (n - 1)
let getBoxCol r c n =
    let box = (r / 3) * 3 + (c / 3)
    243 + box * 9 + (n - 1)

let buildDlxRow r c n id =
    rowInfo.[id] <- { Row = r; Col = c; Num = n }
    
    let n1 = addNode columns.[getPositionCol r c] id
    let n2 = addNode columns.[getRowCol r n] id
    let n3 = addNode columns.[getColCol c n] id
    let n4 = addNode columns.[getBoxCol r c n] id
    
    n1.Right <- n2; n2.Right <- n3; n3.Right <- n4; n4.Right <- n1
    n1.Left <- n4; n2.Left <- n1; n3.Left <- n2; n4.Left <- n3
    
    rowStarts.[id] <- n1

let buildMatrix () =
    let mutable id = 0
    for r in 0 .. 8 do
        for c in 0 .. 8 do
            if puzzle.[r,c] <> 0 then
                buildDlxRow r c puzzle.[r,c] id
                id <- id + 1
            else
                for n in 1 .. 9 do
                    buildDlxRow r c n id
                    id <- id + 1

let cover (c: DlxColumn) =
    c.Right.Left <- c.Left
    c.Left.Right <- c.Right
    
    let mutable row = c.Down
    while row <> (c :> DlxNode) do
        let mutable right = row.Right
        while right <> row do
            right.Down.Up <- right.Up
            right.Up.Down <- right.Down
            right.Column.Size <- right.Column.Size - 1
            right <- right.Right
        row <- row.Down

let uncover (c: DlxColumn) =
    let mutable row = c.Up
    while row <> (c :> DlxNode) do
        let mutable left = row.Left
        while left <> row do
            left.Column.Size <- left.Column.Size + 1
            left.Down.Up <- left
            left.Up.Down <- left
            left <- left.Left
        row <- row.Up
    c.Right.Left <- c
    c.Left.Right <- c

let chooseColumn () =
    let mutable best: DlxColumn = null
    let mutable minSize = Int32.MaxValue
    let mutable curr = root.Right
    while curr <> (root :> DlxNode) do
        let col = curr :?> DlxColumn
        if col.Size < minSize then
            minSize <- col.Size
            best <- col
        curr <- curr.Right
    best

let rec dlxSearch k (solution: int[]) =
    dlxIterations <- dlxIterations + 1
    
    if root.Right = (root :> DlxNode) then true // solved
    else
        let col = chooseColumn()
        if col.Size = 0 then false
        else
            cover col
            let mutable row = col.Down
            let mutable found = false
            
            while row <> (col :> DlxNode) && not found do
                solution.[k] <- row.RowId
                let mutable right = row.Right
                while right <> row do
                    cover right.Column
                    right <- right.Right
                
                if dlxSearch (k + 1) solution then
                    found <- true
                else
                    // Backtrack
                    let mutable left = row.Left
                    while left <> row do
                        uncover left.Column
                        left <- left.Left
                    solution.[k] <- -1
                
                row <- row.Down
                
            uncover col
            found

let coverClues () =
    for r in 0 .. 8 do
        for c in 0 .. 8 do
            if puzzle.[r,c] <> 0 then
                let n = puzzle.[r,c]
                // Find matching rowId
                let mutable found = false
                let mutable id = 0
                while not found && id < 729 do
                   if rowStarts.[id] <> null && rowInfo.[id].Row = r && rowInfo.[id].Col = c && rowInfo.[id].Num = n then
                       found <- true
                       let mutable curr = rowStarts.[id]
                       let start = curr
                       let mutable first = true
                       while first || curr <> start do
                           first <- false
                           cover curr.Column
                           curr <- curr.Right
                   id <- id + 1

let extractSolution (sol: int[]) =
    solutionGrid <- Array2D.copy puzzle
    for i in 0 .. 80 do
        let id = sol.[i]
        if id >= 0 then
            let info = rowInfo.[id]
            solutionGrid.[info.Row, info.Col] <- info.Num

let printGrid (grid: int[,]) =
    for r in 0 .. 8 do
        for c in 0 .. 8 do
            printf "%d " grid.[r,c]
        printfn ""

let readMatrix (file: string) =
    if not (File.Exists file) then
        eprintfn "Error reading %s" file
        false
    else
        let display = if file.StartsWith("/app/Matrices/") then "../" + file.Substring(5) else file
        printfn "%s" display
        let lines = File.ReadAllLines file |> Array.filter (fun l -> not (String.IsNullOrWhiteSpace l) && not (l.StartsWith "#"))
        let mutable r = 0
        for l in lines do
            if r < 9 then
                let parts = l.Split([|' '; '\t'|], StringSplitOptions.RemoveEmptyEntries)
                if parts.Length = 9 then
                    for c in 0 .. 8 do
                        puzzle.[r,c] <- int parts.[c]
                        printf "%d " puzzle.[r,c]
                    printfn ""
                    r <- r + 1
        r = 9

[<EntryPoint>]
let main args =
    let sw = Stopwatch.StartNew()
    
    for arg in args do
        if readMatrix arg then
            printfn "\nPuzzle:"
            printGrid puzzle
            
            dlxIterations <- 0
            initDlxMatrix()
            buildMatrix()
            coverClues()
            
            let solution = Array.create 81 (-1)
            let result = dlxSearch 0 solution
            
            if result then
                extractSolution solution
                printfn "\nPuzzle:"
                printGrid solutionGrid
                printfn "\nSolved in Iterations=%d\n" dlxIterations
            else
                printfn "\nNo solution found"
    
    sw.Stop()
    printfn "Seconds to process %.3f" sw.Elapsed.TotalSeconds
    0

File: Algorithms/DLX/Octave/dlx.m
#!/usr/bin/env octave
% Sudoku Solver - Dancing Links (DLX) Implementation in GNU Octave
% Exact cover problem solved with Algorithm X using Dancing Links
% Uses global arrays for node storage (Octave passes structs by value)

global dlx_iterations;
global nodes;  % Array of all nodes
global columns;  % Array of column headers
global node_count;
global col_count;

% ============================================================================
% NODE MANAGEMENT
% ============================================================================

% Node fields: left, right, up, down, column (index), row_id
% Column fields: size, name (stored separately)

function init_globals()
    global nodes columns node_count col_count;
    nodes = [];
    columns = [];
    node_count = 0;
    col_count = 0;
endfunction

function idx = create_node(row_id)
    global nodes node_count;
    node_count = node_count + 1;
    idx = node_count;
    nodes(idx).left = idx;
    nodes(idx).right = idx;
    nodes(idx).up = idx;
    nodes(idx).down = idx;
    nodes(idx).column = 0;
    nodes(idx).row_id = row_id;
endfunction

function idx = create_column(name)
    global nodes columns node_count col_count;

    % Create node for column header
    node_idx = create_node(-1);

    % Create column entry
    col_count = col_count + 1;
    idx = col_count;
    columns(idx).node = node_idx;
    columns(idx).size = 0;
    columns(idx).name = name;

    nodes(node_idx).column = idx;
endfunction

% ============================================================================
% DLX CORE OPERATIONS
% ============================================================================

function cover_column(c)
    global nodes columns;

    col_node = columns(c).node;

    % Remove column header from header list
    nodes(nodes(col_node).right).left = nodes(col_node).left;
    nodes(nodes(col_node).left).right = nodes(col_node).right;

    % For each row in this column
    i = nodes(col_node).down;
    while i != col_node
        % For each node in this row
        j = nodes(i).right;
        while j != i
            % Remove node from its column
            nodes(nodes(j).down).up = nodes(j).up;
            nodes(nodes(j).up).down = nodes(j).down;
            columns(nodes(j).column).size = columns(nodes(j).column).size - 1;
            j = nodes(j).right;
        endwhile
        i = nodes(i).down;
    endwhile
endfunction

function uncover_column(c)
    global nodes columns;

    col_node = columns(c).node;

    % For each row in this column (reverse order)
    i = nodes(col_node).up;
    while i != col_node
        % For each node in this row (reverse order)
        j = nodes(i).left;
        while j != i
            % Restore node to its column
            columns(nodes(j).column).size = columns(nodes(j).column).size + 1;
            nodes(nodes(j).down).up = j;
            nodes(nodes(j).up).down = j;
            j = nodes(j).left;
        endwhile
        i = nodes(i).up;
    endwhile

    % Restore column header
    nodes(nodes(col_node).right).left = col_node;
    nodes(nodes(col_node).left).right = col_node;
endfunction

function c = choose_column(root)
    global nodes columns;

    c = 0;
    min_size = inf;

    j = nodes(root).right;
    while j != root
        col_idx = nodes(j).column;
        if columns(col_idx).size < min_size
            min_size = columns(col_idx).size;
            c = col_idx;
        endif
        j = nodes(j).right;
    endwhile
endfunction

function [success, solution] = search(root, k, solution)
    global dlx_iterations nodes columns;

    dlx_iterations = dlx_iterations + 1;

    % If matrix is empty, solution found
    if nodes(root).right == root
        success = 1;
        return;
    endif

    % Choose column with minimum size
    c = choose_column(root);

    if c == 0 || columns(c).size == 0
        success = 0;
        return;
    endif

    % Cover this column
    cover_column(c);

    % Try each row in this column
    r = nodes(columns(c).node).down;
    while r != columns(c).node
        % Add row to solution
        solution(k) = nodes(r).row_id;

        % Cover all other columns in this row
        j = nodes(r).right;
        while j != r
            cover_column(nodes(j).column);
            j = nodes(j).right;
        endwhile

        % Recurse
        [success, solution] = search(root, k + 1, solution);
        if success == 1
            return;
        endif

        % Backtrack: uncover columns
        j = nodes(r).left;
        while j != r
            uncover_column(nodes(j).column);
            j = nodes(j).left;
        endwhile

        r = nodes(r).down;
    endwhile

    % Uncover column
    uncover_column(c);

    success = 0;
endfunction

% ============================================================================
% MATRIX CONSTRUCTION
% ============================================================================

function root = build_matrix()
    global nodes columns col_count;

    % Create root
    root = create_column("root");
    root_node = columns(root).node;

    % Create 324 columns
    for i = 1:324
        c = create_column(sprintf("C%d", i-1));
    endfor

    % Link columns in circular list
    % root is at index 1, columns 1-324 are at indices 2-325
    for i = 2:325
        if i == 2
            nodes(columns(i).node).left = root_node;
            nodes(root_node).right = columns(i).node;
        else
            nodes(columns(i).node).left = columns(i-1).node;
            nodes(columns(i-1).node).right = columns(i).node;
        endif

        if i == 325
            nodes(columns(i).node).right = root_node;
            nodes(root_node).left = columns(i).node;
        endif
    endfor
endfunction

function first_node = add_row_to_matrix(row_id, cols)
    global nodes columns;

    row_nodes = [];

    % Create nodes for this row
    for i = 1:length(cols)
        col_idx = cols(i) + 2;  % +2 because root is at index 1, columns start at index 2

        node_idx = create_node(row_id);
        nodes(node_idx).column = col_idx;

        % Insert at end of column
        col_node = columns(col_idx).node;
        nodes(node_idx).up = nodes(col_node).up;
        nodes(node_idx).down = col_node;
        nodes(nodes(col_node).up).down = node_idx;
        nodes(col_node).up = node_idx;
        columns(col_idx).size = columns(col_idx).size + 1;

        row_nodes(end+1) = node_idx;
    endfor

    % Link row nodes horizontally
    if length(row_nodes) > 0
        for i = 1:length(row_nodes)
            if i == 1
                nodes(row_nodes(i)).left = row_nodes(end);
                nodes(row_nodes(end)).right = row_nodes(i);
            else
                nodes(row_nodes(i)).left = row_nodes(i-1);
                nodes(row_nodes(i-1)).right = row_nodes(i);
            endif
        endfor
        first_node = row_nodes(1);
    else
        first_node = 0;
    endif
endfunction

% ============================================================================
% PUZZLE PARSING
% ============================================================================

function [root, puzzle, row_info, row_starts] = parse_matrix(filename)
    init_globals();

    % Read puzzle
    puzzle = zeros(9, 9);
    fid = fopen(filename, 'r');
    if fid == -1
        error('Cannot open file: %s', filename);
    endif

    line_count = 0;
    while ~feof(fid)
        line = fgetl(fid);
        if ischar(line)
            line = strtrim(line);
            if ~isempty(line) && line(1) != '#'
                values = sscanf(line, '%d');
                if length(values) == 9
                    line_count = line_count + 1;
                    puzzle(line_count, :) = values';
                    if line_count == 9
                        break;
                    endif
                endif
            endif
        endif
    endwhile
    fclose(fid);

    % Build matrix
    root = build_matrix();

    % Add rows for all possibilities (not just valid ones)
    % Store row metadata
    row_info = cell(729, 1);
    row_starts = zeros(729, 1);
    row_id_idx = 1;

    for row = 0:8
        for col = 0:8
            if puzzle(row+1, col+1) != 0
                % Cell has a clue - create only one row
                digit = puzzle(row+1, col+1);
                c1 = row * 9 + col;
                c2 = 81 + row * 9 + (digit - 1);
                c3 = 162 + col * 9 + (digit - 1);
                box = floor(row / 3) * 3 + floor(col / 3);
                c4 = 243 + box * 9 + (digit - 1);

                row_id = row * 81 + col * 9 + digit;
                first_node = add_row_to_matrix(row_id, [c1, c2, c3, c4]);
                row_info{row_id_idx} = struct('row', row, 'col', col, 'digit', digit);
                row_starts(row_id_idx) = first_node;
                row_id_idx = row_id_idx + 1;
            else
                % Cell is empty - create rows for all possible values
                for digit = 1:9
                    c1 = row * 9 + col;
                    c2 = 81 + row * 9 + (digit - 1);
                    c3 = 162 + col * 9 + (digit - 1);
                    box = floor(row / 3) * 3 + floor(col / 3);
                    c4 = 243 + box * 9 + (digit - 1);

                    row_id = row * 81 + col * 9 + digit;
                    first_node = add_row_to_matrix(row_id, [c1, c2, c3, c4]);
                    row_info{row_id_idx} = struct('row', row, 'col', col, 'digit', digit);
                    row_starts(row_id_idx) = first_node;
                    row_id_idx = row_id_idx + 1;
                endfor
            endif
        endfor
    endfor
endfunction

% ============================================================================
% CLUE COVERING
% ============================================================================

function cover_clues_func(puzzle, row_info, row_starts)
    global nodes;

    for row = 0:8
        for col = 0:8
            if puzzle(row+1, col+1) != 0
                digit = puzzle(row+1, col+1);

                % Find the row for this clue
                for i = 1:length(row_info)
                    if row_starts(i) != 0 && ~isempty(row_info{i})
                        if row_info{i}.row == row && row_info{i}.col == col && row_info{i}.digit == digit
                            % Cover all columns in this row
                            node = row_starts(i);
                            curr = node;
                            while true
                                cover_column(nodes(curr).column);
                                curr = nodes(curr).right;
                                if curr == node
                                    break;
                                endif
                            endwhile
                            break;
                        endif
                    endif
                endfor
            endif
        endfor
    endfor
endfunction

% ============================================================================
% SOLUTION EXTRACTION
% ============================================================================

function grid = extract_solution(solution, num_rows)
    grid = zeros(9, 9);

    for i = 1:num_rows
        row_id = solution(i);

        if row_id > 0
            digit = mod(row_id, 9);
            if digit == 0
                digit = 9;
            endif

            temp = floor((row_id - digit) / 9);
            col = mod(temp, 9);
            row = floor(temp / 9);

            if row >= 0 && row < 9 && col >= 0 && col < 9
                grid(row+1, col+1) = digit;
            endif
        endif
    endfor
endfunction

% ============================================================================
% I/O
% ============================================================================

function print_puzzle(grid)
    fprintf('\nPuzzle:\n');
    for row = 1:9
        for col = 1:9
            fprintf('%d ', grid(row, col));
        endfor
        fprintf('\n');
    endfor
endfunction

% ============================================================================
% MAIN
% ============================================================================

function main()
    global dlx_iterations columns;

    args = argv();
    start_time = time();

    for i = 1:length(args)
        filename = args{i};
        if length(filename) > 7 && strcmp(filename(end-6:end), '.matrix') == 1
            % Print filename
            display_path = filename;
            if strncmp(filename, '/app/Matrices/', 14) == 1
                display_path = filename(6:end);
                fprintf('../%s\n', display_path);
            else
                fprintf('%s\n', filename);
            endif

            % Parse and build matrix
            [root, puzzle, row_info, row_starts] = parse_matrix(filename);

            % Print initial puzzle
            for row = 1:9
                for col = 1:9
                    fprintf('%d ', puzzle(row, col));
                endfor
                fprintf('\n');
            endfor
            print_puzzle(puzzle);

            % Cover pre-filled clues
            cover_clues_func(puzzle, row_info, row_starts);

            % Solve
            dlx_iterations = 0;
            solution = zeros(81, 1);
            [success, solution] = search(columns(root).node, 1, solution);

            if success == 1
                % Start with original puzzle (includes clues)
                grid = puzzle;
                % Add solution values
                for i = 1:81
                    if solution(i) > 0
                        row_id = solution(i);
                        digit = mod(row_id, 9);
                        if digit == 0
                            digit = 9;
                        endif
                        temp = floor((row_id - digit) / 9);
                        col = mod(temp, 9);
                        row = floor(temp / 9);
                        if row >= 0 && row < 9 && col >= 0 && col < 9
                            grid(row+1, col+1) = digit;
                        endif
                    endif
                endfor
                print_puzzle(grid);
                fprintf('\nSolved in Iterations=%d\n\n', dlx_iterations);
            else
                fprintf('No solution found\n');
            endif
        endif
    endfor

    elapsed = time() - start_time;
    fprintf('Seconds to process %.3f\n', elapsed);
endfunction

main();
